Unwind on array of objects crashes

Hi.
Here is my code I want to run and do writing in Neo4j and I get an error.
I don't know from where it takes "type" field name. I guess the problem is coming from UNWIND on array of objects.
I guess it does not work as expected.

```
sparkVersion = "2.4.7"
scalaVersion = "2.12.12"
neo4jConnectorVersion = "4.0.0"
jdkVersion = "14"
```
```
case class User(id: String, firstName: String, lastName: String, gender: String, birthday: String, email: String, arrayOfObjects: Array[Obj])
case class Obj(objStringField1: String, objStringField2: String)
```
Smth like this:
```
{
  "id": "df25368a-49a1-4843-9d49-dd7de700763e",
  "firstName": "Rolfe",
  "lastName": "Noyce",
  "gender": "MALE",
  "birthday": "2000-12-19T04:00:00.000+04:00",
  "email": "rnoyce1@google.co.uk",
  "arrayOfObjects": [
    {
      "objStringField1": "test1",
      "objStringField2": "test2"
    }
  ]
}
```

```
    users.write
      .format("org.neo4j.spark.DataSource")
      .option("database", "{my-db}")
      .option("url", "bolt://{my-neo4j-server}")
      .option("authentication.basic.username", "{my-username}")
      .option("authentication.basic.password", "{my-password}")
      .mode(SaveMode.Overwrite)
      .option("query",
        """
          |UNWIND $event.arrayOfObjects as obj
          |MERGE (u:User {id: event.id, firstName: event.firstName, lastName: event.lastName, gender: event.gender, birthday: event.birthday, email: event.email, userProp: obj.objStringField1})
          |""".stripMargin)
      .save()
```
It crashes. Here is the stacktrace:
```
Caused by: java.lang.IllegalArgumentException: Field "type" does not exist.
Available fields: objStringField1, objStringField2
    at org.apache.spark.sql.types.StructType.$anonfun$fieldIndex$1(StructType.scala:305)
    at scala.collection.immutable.Map$Map2.getOrElse(Map.scala:222)
    at org.apache.spark.sql.types.StructType.fieldIndex(StructType.scala:303)
    at org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema.fieldIndex(rows.scala:187)
    at org.neo4j.spark.util.Neo4jUtil$.convertFromSpark(Neo4jUtil.scala:171)
    at org.neo4j.spark.util.Neo4jUtil$.$anonfun$convertFromSpark$1(Neo4jUtil.scala:197)
    at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:285)
    at scala.collection.Iterator.foreach(Iterator.scala:943)
    at scala.collection.Iterator.foreach$(Iterator.scala:943)
    at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
    at scala.collection.IterableLike.foreach(IterableLike.scala:74)
    at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
    at org.apache.spark.sql.catalyst.util.ArrayDataIndexedSeq.foreach(ArrayData.scala:205)
    at scala.collection.TraversableLike.map(TraversableLike.scala:285)
    at scala.collection.TraversableLike.map$(TraversableLike.scala:278)
    at org.apache.spark.sql.catalyst.util.ArrayDataIndexedSeq.map(ArrayData.scala:205)
    at org.neo4j.spark.util.Neo4jUtil$.convertFromSpark(Neo4jUtil.scala:197)
    at org.neo4j.spark.service.Neo4jWriteMappingStrategy.$anonfun$query$1(MappingService.scala:125)
    at org.neo4j.spark.service.Neo4jWriteMappingStrategy.$anonfun$query$1$adapted(MappingService.scala:123)
    at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:292)
    at scala.collection.immutable.Range.foreach(Range.scala:158)
    at scala.collection.TraversableLike.flatMap(TraversableLike.scala:292)
    at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:289)
    at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
    at org.neo4j.spark.service.Neo4jWriteMappingStrategy.query(MappingService.scala:123)
    at org.neo4j.spark.service.Neo4jWriteMappingStrategy.query(MappingService.scala:20)
    at org.neo4j.spark.service.MappingService.convert(MappingService.scala:234)
    at org.neo4j.spark.writer.Neo4jDataWriter.write(Neo4jDataWriter.scala:41)
    at org.neo4j.spark.writer.Neo4jDataWriter.write(Neo4jDataWriter.scala:20)
    at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.$anonfun$run$2(WriteToDataSourceV2Exec.scala:118)
    at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1394)
    at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:116)
    at org.apache.spark.sql.execution.datasources.v2.WriteToDataSourceV2Exec.$anonfun$doExecute$2(WriteToDataSourceV2Exec.scala:67)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
    at org.apache.spark.scheduler.Task.run(Task.scala:123)
    at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:411)
    at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
```

--- a/pom.xml
+++ b/pom.xml
@@ -47,7 +47,7 @@
         <driver.version>4.1.1</driver.version>
         <spark.version>2.4.5</spark.version>
         <neo4j.experimental>false</neo4j.experimental>
-        <testcontainers.version>1.12.5</testcontainers.version>
+        <testcontainers.version>1.15.0</testcontainers.version>
         <xbean.version>4.10</xbean.version>
         <cypherdsl.version>2020.0.0</cypherdsl.version>
         <node.version>v12.16.0</node.version>
@@ -302,10 +302,10 @@
                     <artifactId>xbean-asm6-shaded</artifactId>
                 </exclusion>
                 <!-- uncomment for debug -->
-<!--                <exclusion>-->
-<!--                    <groupId>org.slf4j</groupId>-->
-<!--                    <artifactId>slf4j-log4j12</artifactId>-->
-<!--                </exclusion>-->
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>slf4j-log4j12</artifactId>
+                </exclusion>
             </exclusions>
         </dependency>
 
@@ -349,12 +349,17 @@
             <scope>test</scope>
         </dependency>
         <!-- uncomment for debug -->
-<!--        <dependency>-->
-<!--            <groupId>org.slf4j</groupId>-->
-<!--            <artifactId>slf4j-simple</artifactId>-->
-<!--            <version>1.7.30</version>-->
-<!--            <scope>test</scope>-->
-<!--        </dependency>-->
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
+            <version>1.7.30</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
+            <version>2.12.0</version>
+        </dependency>
     </dependencies>
 
     <repositories>

--- a/pom.xml
+++ b/pom.xml
@@ -302,10 +302,10 @@
                     <artifactId>xbean-asm6-shaded</artifactId>
                 </exclusion>
                 <!-- uncomment for debug -->
-                <exclusion>
-                    <groupId>org.slf4j</groupId>
-                    <artifactId>slf4j-log4j12</artifactId>
-                </exclusion>
+<!--                <exclusion>-->
+<!--                    <groupId>org.slf4j</groupId>-->
+<!--                    <artifactId>slf4j-log4j12</artifactId>-->
+<!--                </exclusion>-->
             </exclusions>
         </dependency>
 
@@ -349,12 +349,12 @@
             <scope>test</scope>
         </dependency>
         <!-- uncomment for debug -->
-        <dependency>
-            <groupId>org.slf4j</groupId>
-            <artifactId>slf4j-simple</artifactId>
-            <version>1.7.30</version>
-            <scope>test</scope>
-        </dependency>
+<!--        <dependency>-->
+<!--            <groupId>org.slf4j</groupId>-->
+<!--            <artifactId>slf4j-simple</artifactId>-->
+<!--            <version>1.7.30</version>-->
+<!--            <scope>test</scope>-->
+<!--        </dependency>-->
         <dependency>
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-annotations</artifactId>

--- a/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
+++ b/src/main/scala/org/neo4j/spark/util/Neo4jImplicits.scala
@@ -1,12 +1,13 @@
 package org.neo4j.spark.util
 
 
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+
 import javax.lang.model.SourceVersion
 import org.apache.spark.sql.types.{DataTypes, StructField, StructType}
 import org.neo4j.driver.types.{Entity, Node, Relationship}
 import org.neo4j.spark.service.SchemaService
 import org.apache.spark.sql.sources.{EqualNullSafe, EqualTo, Filter, GreaterThan, GreaterThanOrEqual, In, IsNotNull, IsNull, LessThan, LessThanOrEqual, Not, StringContains, StringEndsWith, StringStartsWith}
-
 import scala.collection.JavaConverters._
 
 object Neo4jImplicits {

--- a/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
+++ b/src/main/scala/org/neo4j/spark/util/Neo4jUtil.scala
@@ -3,7 +3,6 @@ package org.neo4j.spark.util
 import java.time._
 import java.time.format.DateTimeFormatter
 import java.util.Properties
-
 import com.fasterxml.jackson.core.JsonGenerator
 import com.fasterxml.jackson.databind.module.SimpleModule
 import com.fasterxml.jackson.databind.{JsonSerializer, ObjectMapper, SerializerProvider}
@@ -16,7 +15,7 @@ import org.apache.spark.unsafe.types.UTF8String
 import org.neo4j.cypherdsl.core.{Condition, Cypher}
 import org.neo4j.driver.internal._
 import org.neo4j.driver.types.{Entity, Path}
-import org.neo4j.driver.{Session, Transaction, Values}
+import org.neo4j.driver.{Session, Transaction, Value, Values}
 import org.neo4j.spark.service.SchemaService
 import org.neo4j.spark.util.Neo4jImplicits.EntityImplicits
 import org.slf4j.Logger
@@ -168,24 +167,32 @@ object Neo4jUtil {
       convertFromSpark(row)
     }
     case struct: GenericRowWithSchema => {
-      if (struct.fieldIndex("type") == -1) {
-        Values.NULL
+      def toMap(struct: GenericRowWithSchema): Value = {
+        Values.value(
+          struct.schema.fields.map(
+            f => f.name -> Neo4jUtil.convertFromSpark(struct.getAs(f.name), f)
+          ).toMap.asJava)
       }
-      struct.getAs[UTF8String]("type").toString match {
-        case SchemaService.POINT_TYPE_2D => Values.point(struct.getAs[Number]("srid").intValue(),
-          struct.getAs[Number]("x").doubleValue(),
-          struct.getAs[Number]("y").doubleValue())
-        case SchemaService.POINT_TYPE_3D => Values.point(struct.getAs[Number]("srid").intValue(),
-          struct.getAs[Number]("x").doubleValue(),
-          struct.getAs[Number]("y").doubleValue(),
-          struct.getAs[Number]("z").doubleValue())
-        case SchemaService.DURATION_TYPE => Values.isoDuration(struct.getAs[Number]("months").longValue(),
-          struct.getAs[Number]("days").longValue(),
-          struct.getAs[Number]("seconds").longValue(),
-          struct.getAs[Number]("nanoseconds").intValue())
-        case SchemaService.TIME_TYPE_OFFSET => Values.value(OffsetTime.parse(struct.getAs[String]("value")))
-        case SchemaService.TIME_TYPE_LOCAL => Values.value(LocalTime.parse(struct.getAs[String]("value")))
-        case _ => Values.NULL // check if it's correct
+
+      try {
+        struct.getAs[UTF8String]("type").toString match {
+          case SchemaService.POINT_TYPE_2D => Values.point(struct.getAs[Number]("srid").intValue(),
+            struct.getAs[Number]("x").doubleValue(),
+            struct.getAs[Number]("y").doubleValue())
+          case SchemaService.POINT_TYPE_3D => Values.point(struct.getAs[Number]("srid").intValue(),
+            struct.getAs[Number]("x").doubleValue(),
+            struct.getAs[Number]("y").doubleValue(),
+            struct.getAs[Number]("z").doubleValue())
+          case SchemaService.DURATION_TYPE => Values.isoDuration(struct.getAs[Number]("months").longValue(),
+            struct.getAs[Number]("days").longValue(),
+            struct.getAs[Number]("seconds").longValue(),
+            struct.getAs[Number]("nanoseconds").intValue())
+          case SchemaService.TIME_TYPE_OFFSET => Values.value(OffsetTime.parse(struct.getAs[String]("value")))
+          case SchemaService.TIME_TYPE_LOCAL => Values.value(LocalTime.parse(struct.getAs[String]("value")))
+          case _ => toMap(struct)
+        }
+      } catch {
+        case _: Throwable => toMap(struct)
       }
     }
     case unsafeArray: UnsafeArrayData => {

--- a/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
+++ b/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
@@ -1271,8 +1271,6 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
                     |}""".stripMargin
     val df = ss.read.json(Seq(jsonStr).toDS)
 
-    df.show()
-
     df.write
       .format(classOf[DataSource].getName)
       .format("org.neo4j.spark.DataSource")

--- a/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
+++ b/src/test/scala/org/neo4j/spark/DataSourceWriterTSE.scala
@@ -1249,4 +1249,40 @@ class DataSourceWriterTSE extends SparkConnectorScalaBaseTSE {
 
     SparkConnectorScalaSuiteIT.session().run("DROP CONSTRAINT ON (m:Musician) ASSERT (m.name) IS UNIQUE")
   }
+
+  /**
+   * https://github.com/neo4j-contrib/neo4j-spark-connector/issues/265
+   */
+  @Test
+  def `should write a custom map with query method`(): Unit = {
+    val jsonStr = """{
+                    |  "id": "df25368a-49a1-4843-9d49-dd7de700763e",
+                    |  "firstName": "Rolfe",
+                    |  "lastName": "Noyce",
+                    |  "gender": "MALE",
+                    |  "birthday": "2000-12-19T04:00:00.000+04:00",
+                    |  "email": "rnoyce1@google.co.uk",
+                    |  "arrayOfObjects": [
+                    |    {
+                    |      "objStringField1": "test1",
+                    |      "objStringField2": "test2"
+                    |    }
+                    |  ]
+                    |}""".stripMargin
+    val df = ss.read.json(Seq(jsonStr).toDS)
+
+    df.show()
+
+    df.write
+      .format(classOf[DataSource].getName)
+      .format("org.neo4j.spark.DataSource")
+      .option("url", SparkConnectorScalaSuiteIT.server.getBoltUrl)
+      .mode(SaveMode.Overwrite)
+      .option("query",
+        """
+          |UNWIND event.arrayOfObjects as obj
+          |MERGE (u:User {id: event.id, firstName: event.firstName, lastName: event.lastName, gender: event.gender, birthday: event.birthday, email: event.email, userProp: obj.objStringField1})
+          |""".stripMargin)
+      .save()
+  }
 }

